### ISSUE TITLE
do not validate Literal as a type form when indexed in value expression context

### ISSUE THREAD

**Issue by reindervdw-cmi** [astral-sh/ty#2905]
### Summary

When using a variable inside a Literal, ty throws an `invalid-type-form` error. However, this is valid usage of Literal e.g. for dynamically constructing Pydantic classes:

```python
from typing import Literal
import pydantic

my_string = "hello world!"

my_dynamic_class = pydantic.create_model(
    "my_dynamic_class",
    literal_string=Literal[
        my_string
    ],  # This gives an invalid-type-form error on Literal[my_string]
)

print(my_dynamic_class(literal_string="hello world!").literal_string) # Prints "hello world!" without error
```

Tested with pydantic 2.12.5 and ty 0.0.14 on Python 3.14

### Version

0.0.14

**Comment by AlexWaygood** [comment]
Thanks for the report. This is basically the same issue as https://github.com/astral-sh/ty/issues/727: we're doing type-form validation everywhere we see a `Literal` slice, even if it's in a runtime context not intended for static type checkers.

Do you have patterns like this regularly occuring in your codebase? My opinion in general is that it's reasonable for us to emit these errors, because situations like this where you need to dynamically construct type forms for a runtime library are pretty rare. But if this is idiomatic and common for code that uses pydantic, then that might change my opinion.

**Comment by carljm** [comment]
It sort of seems like Pydantic wants these to be type forms, and I would suspect may eventually use `TypeForm` (PEP 747) to annotate them as such. In which case it will be correct to error here, at least according to [the current spec](https://typing.python.org/en/latest/spec/literal.html#legal-and-illegal-parameterizations) for `Literal`.

But regardless I do think we should avoid doing this validation when `Literal` is indexed in a non-type-expression context.

**Comment by reindervdw-cmi** [comment]
@AlexWaygood I suspect this is an uncommon use-case. My specific use-case is the runtime creation of a response model for a query to a Large Language Model. I need an LLM to answer a set of questions only known at runtime, and I use the literal strings to enforce that the response contains the question, verbatim, to ensure proper matching between question/response.

**Comment by AlexWaygood** [comment]
I'll close this issue in favour of #727, as this is the same issue just in a slightly different context, and will add a comment to that issue noting that we had it reported here with the `Literal` context
